examples/nmf/script.jl

```diff
@@ -0,0 +1,347 @@
+# &\min_{G \in \mathbb{R}^{m \times n}} \sum_{i = 1}^n \operatorname{OT}_{\varepsilon}^*(X_i, G_i) + \rho_1 \sum_{i = 1}^{n} E^*(-(D^\top G)_i / \rho_1). 
+# \end{aligned}
+# ```
+# The semi-dual of entropy-regularised optimal transport loss, $\operatorname{OT}_{\varepsilon}$, is implemented as `Dual.ot_entropic_semidual` and its gradient can be computed by `Dual.ot_entropic_semidual_grad`. $E^*$ turns out to be `logsumexp`, for which we implement a wrapper function:
+#
+function E_star(x; dims=1)
+    return logsumexp(x; dims=dims)
+end;
+# 
+# The gradient of `logsumexp` is `softmax`, so we define also its gradient:
+#
+function E_star_grad(x; dims=1)
+    return NNlib.softmax(x; dims=1)
+end;
+# 
+# Thus, for each problem we may define the dual objective and its gradient. We note that `ot_entropic_semidual(_grad)` automatically broadcasts along columns of its input. There is therefore no need to make multiple calls to the function, thus allowing for more efficient evaluation.
+#
+function dual_obj_weights(X, K, ε, D, G, ρ1)
+    return sum(Dual.ot_entropic_semidual(X, G, ε, K)) + ρ1 * sum(E_star(-D' * G / ρ1))
+end
+function dual_obj_weights_grad!(∇, X, K, ε, D, G, ρ1)
+    return ∇ .= Dual.ot_entropic_semidual_grad(X, G, ε, K) - D * E_star_grad(-D' * G / ρ1)
+end
+function dual_obj_dict(X, K, ε, Λ, G, ρ2)
+    return sum(Dual.ot_entropic_semidual(X, G, ε, K)) + ρ2 * sum(E_star(-G * Λ' / ρ2))
+end
+function dual_obj_dict_grad!(∇, X, K, ε, Λ, G, ρ2)
+    return ∇ .= Dual.ot_entropic_semidual_grad(X, G, ε, K) - E_star_grad(-G * Λ' / ρ2) * Λ
+end;
+# 
+# The only remaining part of Wasserstein-NMF to implement is the conversion at optimality from the dual variable $G$ to the primal variables $D, \Lambda$. From the results of Theorems 3 and 4 in Rolet et al., we have for $\Lambda$:
+# ```math
+# \Lambda_i = \operatorname{softmax}((-D^\top G)_i / \rho_1),
+# ```
+# and we have for $D$:
+# ```math
+# D_i = \operatorname{softmax}((-G \Lambda^\top)_i / \rho_2). 
+# ```
+#
+function getprimal_weights(D, G, ρ1)
+    return NNlib.softmax(-D' * G / ρ1; dims=1)
+end
+function getprimal_dict(Λ, G, ρ2)
+    return NNlib.softmax(-G * Λ' / ρ2; dims=1)
+end;
+# 
+# We can now implement functions `solve_weights` and `solve_dict` that solve the respective dual problems for the next iterates of `Λ` and `D`. 
+#
+function solve_weights(X, K, ε, D, ρ1; alg, options)
+    opt = optimize(
+        g -> dual_obj_weights(X, K, ε, D, g, ρ1),
+        (∇, g) -> dual_obj_weights_grad!(∇, X, K, ε, D, g, ρ1),
+        zero.(X),
+        alg,
+        options,
+    )
+    return getprimal_weights(D, Optim.minimizer(opt), ρ1)
+end
+function solve_dict(X, K, ε, Λ, ρ2; alg, options)
+    opt = optimize(
+        g -> dual_obj_dict(X, K, ε, Λ, g, ρ2),
+        (∇, g) -> dual_obj_dict_grad!(∇, X, K, ε, Λ, g, ρ2),
+        zero.(X),
+        alg,
+        options;
+        autodiff=:forward,
```

Comment:
This is the default in Optim, isn't it?

Reply:
BTW, also in the other example, intuitively I would assume that forward mode AD is slower than reverse mode AD.

Reply:
Yes, reverse diff is faster for this. Looks like Optim.jl doesn't allow `autodiff=:reverse` but I'll add this manually.

Reply:
> This is the default in Optim, isn't it?
No, I think the default is to use finite differences. But not necessary when the gradient is already provided. Fixed